src/internal/easee/signalr.go

```diff
@@ -0,0 +1,451 @@
+package easee
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/futurehomeno/cliffhanger/root"
+	libsignalr "github.com/philippseith/signalr"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/futurehomeno/edge-easee-adapter/internal/config"
+	"github.com/futurehomeno/edge-easee-adapter/internal/signalr"
+)
+
+const (
+	signalRURI = "/hubs/chargers"
+)
+
+// SignalRManager is the interface for the Easee signalR manager.
+// It manages the signalR connection and the chargers that are connected to it.
+type SignalRManager interface {
+	root.Service
+
+	// Register registers a charger to be managed.
+	Register(chargerID string, cache ObservationCache, callbacks map[ObservationID]func())
+	// Unregister unregisters a charger from being managed.
+	Unregister(chargerID string)
+}
+
+type signalRManager struct {
+	mu       sync.RWMutex
+	client   signalr.Client
+	receiver *SignalRReceiver
+	chargers map[string]chargerItem
+	done     chan struct{}
+}
+
+func NewSignalRManager(client signalr.Client, receiver *SignalRReceiver) SignalRManager {
+	return &signalRManager{
+		client:   client,
+		receiver: receiver,
+		chargers: make(map[string]chargerItem),
+		done:     make(chan struct{}),
+	}
+}
+
+func (m *signalRManager) Start() error {
+	go m.run()
+	go m.handleObservations()
+
+	return nil
+}
+
+func (m *signalRManager) Stop() error {
+	close(m.done)
+	m.client.Close()
+
+	return nil
+}
+
+func (m *signalRManager) Register(chargerID string, cache ObservationCache, callbacks map[ObservationID]func()) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.chargers[chargerID]; ok {
+		return
+	}
+
+	m.chargers[chargerID] = chargerItem{
+		cache:     cache,
+		callbacks: callbacks,
+	}
+
+	if m.client.Connected() {
+		if err := m.client.SubscribeCharger(chargerID); err != nil {
+			log.WithError(err).Error("failed to subscribe charger: ", chargerID)
+
+			return
+		}
+
+		cache.setConnected(true)
+	}
+}
+
+func (m *signalRManager) Unregister(chargerID string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.chargers[chargerID]; !ok {
+		return
+	}
+
+	delete(m.chargers, chargerID)
+
+	if err := m.client.UnsubscribeCharger(chargerID); err != nil {
+		log.WithError(err).Error("failed to unsubscribe charger: ", chargerID)
+	}
+}
+
+func (m *signalRManager) run() {
+	ch := m.client.ObserveState()
+
+	m.client.Start()
+
+	for {
+		select {
+		case <-m.done:
+			return
+		case state := <-ch:
+			m.mu.RLock()
+
+			if state == signalr.Disconnected {
+				for _, item := range m.chargers {
+					item.cache.setConnected(false)
+				}
+
+				m.mu.RUnlock()
+
+				continue
+			}
+
+			for chargerID, charger := range m.chargers {
+				charger.cache.setConnected(true)
```

Comment:
Why this is before the error? Is there any reason for it?

Reply:
We discussed that on PM - there's no ideal solution here.